R/Editor.R

```diff
@@ -2,109 +2,204 @@ editor_ui <- function(id, x) {
+          )
+        }
+        NULL
+      })
     })
 
+    blocks_queue_r(names(template_card)) # Initialize the already rendered blocks
+
     shiny::observeEvent(input$add_block, {
-      # because only new names will be called (see blocks_new)
-      new_name <- utils::tail(
-        make.unique(
-          c(
-            blocks_called(),
-            "block"
-          )
-        ),
-        1
-      )
-      x_reactive(
-        modifyList(x_reactive(), stats::setNames(list(""), new_name)) # Preserve attributes
-      )
+      new_name <- utils::tail(make.unique(c(names(template_card), names(blocks_new_value), "block"), sep = "_"), 1)
+      blocks_new_value[[new_name]] <- ""
+      blocks_queue_r(new_name)
     })
+
+    blocks_input_names
   })
 }
 
-#' @export
-editor_ui.default <- function(id, x) {
-  toHTML(x)
+ui_edit_button <- function(id) {
+  shiny::actionLink(
+    inputId = NS(id, "button"),
+    class = "btn btn-primary btn-sm float-end p-3",
+    label = NULL,
+    title = "Edit card",
+    icon = shiny::icon("edit")
+  )
 }
 
-#' @export
-editor_srv.default <- function(id, x, x_reactive) {
-  shiny::moduleServer(id, function(input, output, session) {
-    x_reactive
-  })
-}
+srv_edit_button <- function(id, card_r, reporter) {
+  moduleServer(id, function(input, output, session) {
+    new_card_r <- shiny::reactiveVal()
+    block_input_names_r <- shiny::reactiveVal()
 
-#' @export
-editor_ui.character <- function(id, x) {
-  ns <- shiny::NS(id)
-  shiny::textAreaInput(ns("content"), label = NULL, value = x)
-}
+    shiny::observeEvent(input$button, {
+      template_card <- card_r()
+      names(template_card) <- make.unique(rep("block", length(template_card)), sep = "_")
+      new_card_r(template_card)
 
-#' @export
-editor_srv.character <- function(id, x, x_reactive) {
-  shiny::moduleServer(id, function(input, output, session) {
-    debounced_content <- shiny::debounce(shiny::reactive(input$content), millis = 10000)
+      random_editor_id <- sprintf("editor_%s", rlang::hash(list(rnorm(1), Sys.time())))
+      shiny::showModal(
+        shiny::modalDialog(
+          title = tags$span(
+            class = "edit_title_container",
+            "Editing Card:",
+            shiny::uiOutput(session$ns("title")),
+            shiny::actionLink(session$ns("edit_title"), label = "(edit title)", class = "text-muted")
+          ),
+          size = "l",
+          easyClose = TRUE,
+          shiny::tagList(
+            ui_report_document_editor(session$ns(random_editor_id), x = template_card),
+            shiny::uiOutput(session$ns("add_text_element_button_ui"))
+          ),
+          footer = shiny::tagList(
+            shiny::actionButton(session$ns("edit_save"), label = "Save"),
+            shiny::modalButton("Close")
+          )
+        )
+      )
 
-    shiny::observeEvent(debounced_content(), {
-      x_reactive(debounced_content())
+      result <- srv_report_document_editor(random_editor_id, template_card)
```

Comment:
New server instance per button click is not good. Maybe show ui n-times but call server only once?

Reply:
I'll try to find a way to better manage the data. I'm not a fan of the previous design and this is not ideal, maybe there's a middle ground!

Reply:
`srv_report_document_editor` can be called once as it gets `template_card` (reactiveVal). Let's call `ui/srv_report_document_editor` with const id, like "editor". UI can be rerendered many times but calling server multiple times (registering observers) can cause unexpected bugs.